stripe/src/test/java/com/stripe/android/StripePaymentAuthTest.kt

```diff
@@ -102,44 +114,160 @@ class StripePaymentAuthTest {
+                    data
+                )
+            ).thenReturn(null)
+
+            val stripe = createStripe()
+            stripe.onPaymentResult(
+                StripePaymentController.PAYMENT_REQUEST_CODE,
+                data,
+                callback = paymentCallback
+            )
+
+            idleLooper()
+
+            verify(paymentCallback).onError(isA<InvalidRequestException>())
+        }
 
     @Test
-    fun onSetupResult_whenShouldHandleResultIsTrue_shouldCallHandleResult() {
-        val data = Intent()
-        whenever(
-            paymentController.shouldHandleSetupResult(
+    fun onSetupResult_whenShouldHandleResultAndControllerReturnsCorrectResult_shouldGetCorrectResult() =
+        testDispatcher.runBlockingTest {
+            val data = Intent()
+            val result = mock<SetupIntentResult>()
+            whenever(
+                paymentController.shouldHandleSetupResult(
+                    StripePaymentController.SETUP_REQUEST_CODE,
+                    data
+                )
+            ).thenReturn(true)
+            whenever(
+                paymentController.getSetupIntentResult(
+                    data
+                )
+            ).thenReturn(result)
+
+            val stripe = createStripe()
+            stripe.onSetupResult(
                 StripePaymentController.SETUP_REQUEST_CODE,
-                data
+                data,
+                callback = setupCallback
             )
-        ).thenReturn(true)
 
-        val stripe = createStripe()
-        stripe.onSetupResult(
-            StripePaymentController.SETUP_REQUEST_CODE,
-            data,
-            callback = setupCallback
-        )
+            idleLooper()
+            verify(setupCallback).onSuccess(result)
+        }
 
-        verify(paymentController).handleSetupResult(data, setupCallback)
-    }
+    @Test
+    fun onSetupResult_whenShouldHandleResultAndControllerReturnsNull_shouldThrowException() =
+        testDispatcher.runBlockingTest {
+            val data = mock<Intent>()
+            whenever(
+                paymentController.shouldHandleSetupResult(
+                    StripePaymentController.SETUP_REQUEST_CODE,
+                    data
+                )
+            ).thenReturn(true)
+            whenever(
+                paymentController.getSetupIntentResult(
+                    data
+                )
+            ).thenReturn(null)
+
+            val stripe = createStripe()
+            stripe.onSetupResult(
+                StripePaymentController.SETUP_REQUEST_CODE,
+                data,
+                callback = setupCallback
+            )
+
+            idleLooper()
+
+            verify(setupCallback).onError(isA<InvalidRequestException>())
+        }
+
+    @Test
+    fun onAuthenticateSourceResult_whenControllerReturnsCorrectResult_shouldGetCorrectResult() =
+        testDispatcher.runBlockingTest {
+            val result = mock<Source>()
```

Comment:
We have `SourceFixtures` if you need a `Source`

Reply:
Here since `paymentController` is mocked, I think it's fine to mock whatever it returns also, since we're only verifying the returned object gets correctly passed to the callback